### PR TITLE
feat: `KeyRange` reads into "temporary" memory, remove `StateSlots` and "state" memory

### LINKED ISSUE #207: Refactor memory to just be memory
Currently we have two memories (state memory and temp memory) these should be unified into one memory with `Vec<Word>`.

### LINKED ISSUE #208: Remove `State` and `StateLen`
Now that programs will output their stack / memory to the next program then we no longer have a purpose for `State` and `StateLen` ops.

### PR DESCRIPTION
**Pull request by mitchmindtree**
This PR is based off the experimental `develop` branch (#212) which includes work toward the new generalized predicate program DAG model.

We remove "state memory" (which was originally used to provide values read from state-read programs to following state-read programs and constraints) in favour of having the `KeyRange` ops read directly into regular "temporary" memory.

Upon program completion, a "parent" node's stack and memory are used to initialize the stack and memory of any child programs, so the idea is that we no longer need this extra side-channel of "state memory" for passing values between programs.

The main difference is that temporary memory is flat. Under this PR, a `KeyRange` op first involves writing the absolute memory address and length for each value, followed by writing the values themselves. The `KeyRange` tests are a bit hairier as a result, but hopefully give an idea of how the behaviour has changed.

Closes #207 
Closes #208

### CODE REVIEW

**Comment by otrho** [comment]
Cool.  If there is going to be a single memory and it ends up being the primary way that the programs pass data along to each other, should we keep a `TRUNC` or `FREE` instruction so that we only pass along actually useful data and can throw away temporary stuff?

**Comment by mitchmindtree** [comment]
> should we keep a TRUNC or FREE instruction so that we only pass along actually useful data and can throw away temporary stuff?

looks like @freesig is a step ahead of us and landed `FREE` a couple of commits ago: https://github.com/essential-contributions/essential-base/blob/e3e169311803e2298e7f53ce89f0fc15441a5525/crates/asm-spec/asm.yml#L530

File: crates/asm-spec/asm.yml [hunk 1]
```diff
@@ -347,54 +347,11 @@ StateRead:
...
-                Returns the length of the state value.
-              panics:
-                - slot_ix is out of range.
-                - delta is not a boolean.
-              stack_in: [slot_ix, delta]
-              stack_out: [len]
-
-            NumSlots:
+            DecisionVarSlots:
               opcode: 0x3B
```
Comment (freesig) [inline crates/asm-spec/asm.yml:351]: I think this and the following byte numbers need to be adjusted. We should have a test for this somehow

File: crates/asm-spec/asm.yml [hunk 2]
```diff
@@ -347,54 +347,11 @@ StateRead:
...
-                `which_slots`:
-                
-                - `0` for decision vars.
-                
-                - `1` for pre state slots.
-                
-                - `2` for post state slots.
-              stack_in: [which_slots]
+              description: Get the number of decision var slots.
+              stack_in: []
```
Comment (freesig) [inline crates/asm-spec/asm.yml:354]:
```suggestion
```

File: crates/check/tests/solution.rs [hunk 3]
```diff
@@ -391,7 +391,7 @@ async fn predicate_graph_state_read() {
             // Push the length, num keys to read and slot index for the `KeyRange` op.
             PUSH(4),
             PUSH(1),
-            PUSH(0),
+            // PUSH(0),
```
Comment (freesig) [inline crates/check/tests/solution.rs:394]: ?

File: crates/state-read-vm/src/state_read.rs [hunk 4]
```diff
@@ -141,12 +142,27 @@ where
...
+    for value in values {
+        let value_len = Word::try_from(value.len()).map_err(|_| MemoryError::Overflow)?;
+        // Write the [index, len] pair.
+        vm.memory.store_range(mem_addr, &[value_addr, value_len])?;
+        // Write the value.
+        vm.memory.store_range(value_addr, &value)?;
+        // No need to check addition here as `store_range` would have failed.
+        value_addr += value_len;
+        mem_addr += 2;
+    }
```
Comment (freesig) [inline crates/state-read-vm/src/state_read.rs:166]: Nice. Maybe we should document this layout somewhere?

**Review by freesig — COMMENTED** [review]
Nice one. Lots simpler. Just a few things

File: crates/asm-spec/asm.yml (hunk 1 above)
Comment (mitchmindtree) [reply crates/asm-spec/asm.yml]: Good spot!

File: crates/check/tests/solution.rs (hunk 3 above)
Comment (mitchmindtree) [reply crates/check/tests/solution.rs]: Ah this was no longer necessary as the following `ALOC` adds the mem_addr onto the stack for us - removed this and improved the comments.

File: crates/state-read-vm/src/state_read.rs (hunk 4 above)
Comment (mitchmindtree) [reply crates/state-read-vm/src/state_read.rs:166]: This is documented in the updated asm.yml docs for the KeyRange ops, but I've also added a small example just to make it a bit clearer.

**Review by freesig — APPROVED** [review]
Nice one. Huge relief to have the memory unified

**Pull request merged**